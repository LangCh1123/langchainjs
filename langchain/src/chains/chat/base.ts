// import { BaseChain, ChainValues } from "../base.js";
// import { BaseChatModel, ChatMessage, Role } from "../../chat_models/base.js";
// import { ChatMemory } from "../../memory/chat_memory.js";
// // import { BasePromptTemplate } from "../../prompts/index.js";
// import { Document } from "../../document.js";
//
// export interface ChatChainInput {
//   humanPrefix: Role;
//   aiPrefix: Role;
// }
//
// function getDefaultStartMessages() {
//   return [
//     {
//       role: "system",
//       text: `You are a chatbot optimized for question answering.
//         Your job is to answer the most recent user question based
//         ONLY on the information they have told you before.
//         Do NOT use other information than what is provided in previous
//         messages by the human.`,
//     },
//   ];
// }
//
// export class ChatQAChain extends BaseChain implements ChatChainInput {
//   humanPrefix: Role = "user";
//
//   aiPrefix: Role = "assistant";
//
//   declare memory: ChatMemory;
//
//   model: BaseChatModel;
//
//   // prompt: BasePromptTemplate;
//
//   questionKey = "question";
//
//   documentsKey = "input_documents";
//
//   outputKey = "response";
//
//   starterMessages: ChatMessage[] = [];
//
//   get inputKeys() {
//     return [this.questionKey, this.documentsKey];
//   }
//
//   get outputKeys() {
//     return [this.outputKey];
//   }
//
//   constructor(fields: {
//     model: BaseChatModel;
//     // prompt: BasePromptTemplate;
//     humanPrefix?: Role;
//     aiPrefix?: Role;
//     starterMessages?: ChatMessage[];
//     memory?: ChatMemory;
//   }) {
//     super();
//     this.model = fields.model;
//     // this.prompt = fields.prompt;
//     this.humanPrefix = fields.humanPrefix ?? this.humanPrefix;
//     this.aiPrefix = fields.aiPrefix ?? this.aiPrefix;
//     this.starterMessages = fields.starterMessages ?? getDefaultStartMessages();
//     this.memory =
//       fields.memory ??
//       new ChatMemory({
//         humanPrefix: this.humanPrefix,
//         aiPrefix: this.aiPrefix,
//         messages: this.starterMessages,
//         inputKey: this.questionKey,
//         outputKey: this.outputKey,
//       });
//   }
//
//   static fromModel(model: BaseChatModel, starterMessages: ChatMessage[]) {
//     return new ChatQAChain({ model, starterMessages });
//   }
//
//   async _call(values: ChainValues): Promise<ChainValues> {
//     const newMessage: ChatMessage = {
//       role: this.humanPrefix,
//       text: values[this.questionKey],
//     };
//     const docs = values[this.documentsKey];
//     const docMessages: ChatMessage[] = docs.map((doc: Document) => ({
//       role: this.humanPrefix,
//       text: doc.pageContent,
//     }));
//     const messages = [...this.starterMessages, ...docMessages, newMessage];
//     const output = await this.model.call(messages);
//     return { [this.outputKey]: output };
//   }
//
//   serialize(): SerializedChatQAChain {
//     return {
//       _type: "chat-qa",
//       // TODO: serialize the rest of the fields
//       // model: this.model.serialize(),
//       // prompt: this.prompt.serialize(),
//       // humanPrefix: this.humanPrefix,
//       // aiPrefix: this.aiPrefix,
//     };
//   }
//
//   _chainType() {
//     return "chat-qa" as const;
//   }
// }
//
// export interface SerializedChatQAChain {
//   _type: "chat-qa";
// }
