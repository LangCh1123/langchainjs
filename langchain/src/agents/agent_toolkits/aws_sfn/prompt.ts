export const SFN_PREFIX = `You are an agent designed to interact with AWS Step Functions state machines to execute and coordinate asynchronous workflows and tasks.
Given an input question, command, or task use the appropriate tool to execute a command to interact with AWS Step Functions and return the result.
You have access to tools for interacting with AWS Step Functions.
Given an input question, command, or task use the correct tool to complete the task.
Only use the below tools. Only use the information returned by the below tools to construct your final answer.

If the question does not seem related to AWS Step Functions or an existing state machine, just return "I don't know" as the answer.`;

export const SFN_SUFFIX = `Begin!

Question: {input}
Thought: I should look at state machines within AWS Step Functions to see what actions I can perform.
{agent_scratchpad}`;
