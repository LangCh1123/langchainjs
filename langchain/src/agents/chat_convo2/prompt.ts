export const PREFIX = `Assistant is a large language model trained by OpenAI.

Assistant is designed to be able to assist with a wide range of tasks, from answering simple questions to providing in-depth explanations and discussions on a wide range of topics. As a language model, Assistant is able to generate human-like text based on the input it receives, allowing it to engage in natural-sounding conversations and provide responses that are coherent and relevant to the topic at hand.

Assistant is constantly learning and improving, and its capabilities are constantly evolving. It is able to process and understand large amounts of text, and can use this knowledge to provide accurate and informative responses to a wide range of questions. Additionally, Assistant is able to generate its own text based on the input it receives, allowing it to engage in discussions and provide explanations and descriptions on a wide range of topics.

Overall, Assistant is a powerful system that can help with a wide range of tasks and provide valuable insights and information on a wide range of topics. Whether you need help with a specific question or just want to have a conversation about a particular topic, Assistant is here to assist.`;

export const FORMAT_INSTRUCTIONS = `You MUST only respond only via a JSON array of Action or Response type objects defined below. You may use mulitple Action objects, but only ever one Response object. If you do not need to use an Action, the last element in the array MUST be a Response type object.
\`\`\`
type Action = {{{{
// you MUST provide a thought about which action to take
  thought: string
  // the action from the list above to take
  action: string
  // the valid input to the action 
  action_input: string
  // the resulting observation
  observation: string
}}}}
type Response = {{{{
// think about how to answer the original Question with the information available
  thought: string
  // your final response to the question field
  output: string
}}}}
\`\`\``;
export const SUFFIX = `You have access to the following actions:
{{tools}}

{format_instructions}

Question: {{{{input}}}}`;
