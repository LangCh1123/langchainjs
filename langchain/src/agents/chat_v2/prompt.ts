import { AgentScratchPadChatPromptTemplate } from "../../agents/schema.js";
import { AgentActionOutputParser } from "../../agents/types.js";
import { BasePromptTemplate } from "../../prompts/base.js";
import {
  HumanMessagePromptTemplate,
  SystemMessagePromptTemplate,
} from "../../prompts/chat.js";
import { AgentAction, AgentFinish } from "../../schema/index.js";
import { Tool } from "../../tools/base.js";

export const PREFIX = `Answer the following questions as best you can. You have access to the following tools:`;
export const FORMAT_INSTRUCTIONS = `The way you use the tools is by specifying a json blob, denoted below by $JSON_BLOB
Specifically, this $JSON_BLOB should have a "action" key (with the name of the tool to use) and a "action_input" key (with the input to the tool going here). 
The $JSON_BLOB should only contain a SINGLE action, do NOT return a list of multiple actions. Here is an example of a valid $JSON_BLOB:

\`\`\`
{{
  "action": "calculator",
  "action_input": "1 + 2"
}}
\`\`\`

ALWAYS use the following format:

Question: the input question you must answer
Thought: you should always think about what to do
Action: 
\`\`\`
$JSON_BLOB
\`\`\`
Observation: the result of the action
... (this Thought/Action/Observation can repeat N times)
Thought: I now know the final answer
Final Answer: the final answer to the original input question`;
export const SUFFIX = `Begin! Reminder to always use the exact characters \`Final Answer\` when responding.`;

export const createPrompt = ({
  tools,
  prefix = PREFIX,
  suffix = SUFFIX,
  formatInstructions = FORMAT_INSTRUCTIONS,
  inputVariables,
}: {
  tools: Tool[];
  prefix: string;
  suffix: string;
  formatInstructions: string;
  inputVariables?: string[];
}): BasePromptTemplate => {
  const toolSummaries = tools
    .map((tool) => `${tool.name}: ${tool.description}`)
    .join("\n");

  const toolNames = tools.map((tool) => tool.name).join(", ");

  const formatInstructionsWithTools = formatInstructions.replace(
    "{tool_names}",
    toolNames
  );

  const template = [
    prefix,
    toolSummaries,
    formatInstructionsWithTools,
    suffix,
  ].join("\n\n");
  const messages = [
    SystemMessagePromptTemplate.fromTemplate(template),
    HumanMessagePromptTemplate.fromTemplate("{input}\n\n{agent_scratchpad}"),
  ];

  return new AgentScratchPadChatPromptTemplate({
    promptMessages: messages,
    inputVariables: inputVariables ?? ["input", "intermediate_steps"],
  });
};

export class ChatOutputParser extends AgentActionOutputParser {
  async parse(text: string): Promise<AgentAction | AgentFinish> {
    if (text.includes("Final Answer:")) {
      const output = text.split("Final Answer:").pop()?.trim() || "";

      // Return values is generally always a dictionary with a single `output` key
      // It is not recommended to try anything else at the moment :)
      return { returnValues: { output }, log: text };
    }

    try {
      const [, action] = text.split("```");
      const response = JSON.parse(action.trim());
      const agentAction: AgentAction = {
        tool: response.action,
        toolInput: response.action_input,
        log: text,
      };

      return agentAction;
    } catch (error) {
      throw new Error(`Could not parse LLM output: ${text}`);
    }
  }

  getFormatInstructions(): string {
    throw new Error("_type not implemented");
  }
}
