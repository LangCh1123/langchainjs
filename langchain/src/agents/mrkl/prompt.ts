import { AI_PROMPT } from "@anthropic-ai/sdk";
import { BaseLanguageModel } from "../../base_language/index.js";

export const PREFIX = `Answer the following questions as best you can. You have access to the following tools:`;
export const FORMAT_INSTRUCTIONS = `Use the following format in your response:

Question: the input question you must answer
Thought: you should always think about what to do
Action: the action to take, should be one of [{tool_names}]
Action Input: the input to the action
Observation: the result of the action
... (this Thought/Action/Action Input/Observation can repeat N times)
Thought: I now know the final answer
Final Answer: the final answer to the original input question`;
export const getSuffixForLLM = (llm?: BaseLanguageModel) => {
  switch (llm?._llmType()) {
    case "anthropic":
      return `Begin!

Question: {input}
${AI_PROMPT} Thought: {agent_scratchpad}`;
    default:
      return `Begin!

Question: {input}
Thought:{agent_scratchpad}`;
  }
};
