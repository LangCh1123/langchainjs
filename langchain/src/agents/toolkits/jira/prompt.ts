export const JIRA_ISSUE_CREATE_PROMPT = `
This tool is a wrapper around jira.js's Jira createIssue API, useful when you need to create a Jira issue. 
The input to this tool is a object specifying the fields of the Jira issue, and will be passed into jira.js's Jira \`createIssue\` function.
For example, to create a low priority task called "test issue" with description "test description", you would pass in the following object: 
{"summary": "test issue", "description": "test description", "issuetype": {"name": "Task"}, "priority": {"name": "Low"}}
`;

export const JIRA_GET_ALL_PROJECTS_PROMPT = `
This tool is a wrapper around jira.js's Jira project API, 
useful when you need to fetch all the projects the user has access to, find out how many projects there are, or as an intermediary step that involves searching by projects. 
there is no input to this tool.
`;

export const JIRA_JQL_PROMPT = `
This tool is a wrapper around jira.js's Jira jql API, useful when you need to search for Jira issues.
The input to this tool is a JQL query string, and will be passed into jira.js's Jira \`jql\` function,
For example, to find all the issues in project "Test" assigned to the me, you would pass in the following string:
project = Test AND assignee = currentUser()
or to find issues with summaries that contain the word "test", you would pass in the following string:
summary ~ 'test'
`;

export const JIRA_CATCH_ALL_PROMPT = `
This tool is a wrapper around jira.js's Jira API for the Version3Client.
There are other dedicated tools for fetching all projects, and creating and searching for issues, 
use this tool if you need to perform any other actions allowed by the jira.js Jira API.
The input to this tool is a dictionary specifying a function from jira.js's Jira API, 
as well as a list of arguments and dictionary of keyword arguments to pass into the function.
For example, to get the email of a user with accountId 'foo' you would pass in the following object:
{class: "Users", function: "getUserEmail", args: {accountId: "foo"}}
or to find out how many groups are in the Jira instance, you would pass in the following string:
{"class": "Groups", function: "findGroups", args: {}}
For more information on the Jira API, refer to https://mrrefactoring.github.io/jira.js/modules/Version3.html
`;
