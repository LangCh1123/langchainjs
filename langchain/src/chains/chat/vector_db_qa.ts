// import { BaseChatModel, BaseChatMessage } from "../../chat_models/base.js";
// import { VectorStore } from "../../vectorstores/base.js";
// import { BaseChain, ChainValues } from "../base.js";
// import { ChatQAChain } from "./base.js";
//
// export class ChatChatVectorDBQAChain extends BaseChain {
//   qaChain: ChatQAChain;
//
//   inputKey = "question";
//
//   outputKey = "response";
//
//   vectorStore: VectorStore;
//
//   k = 5;
//
//   get inputKeys() {
//     return [this.inputKey];
//   }
//
//   get outputKeys() {
//     return [this.outputKey];
//   }
//
//   constructor(fields: {
//     vectorStore: VectorStore;
//     qaChain: ChatQAChain;
//     inputKey?: string;
//     outputKey?: string;
//     k?: number;
//   }) {
//     super();
//     this.vectorStore = fields.vectorStore;
//     this.qaChain = fields.qaChain;
//     this.inputKey = fields.inputKey ?? this.inputKey;
//     this.outputKey = fields.outputKey ?? this.outputKey;
//     this.k = fields.k ?? this.k;
//   }
//
//   static fromModel(
//     model: BaseChatModel,
//     vectorStore: VectorStore,
//     starterMessages?: BaseChatMessage[]
//   ) {
//     const qaChain = new ChatQAChain({ model, starterMessages });
//     return new ChatChatVectorDBQAChain({ vectorStore, qaChain });
//   }
//
//   async _call(values: ChainValues): Promise<ChainValues> {
//     const question = values[this.inputKey];
//     const docs = await this.vectorStore.similaritySearch(question, this.k);
//     const result = await this.qaChain.call({
//       [this.qaChain.questionKey]: question,
//       [this.qaChain.documentsKey]: docs,
//     });
//     return { [this.outputKey]: result[this.qaChain.outputKey] };
//   }
//
//   serialize(): SerializedChatQAChain {
//     return {
//       _type: "chat-qa",
//       // TODO: serialize the rest of the fields
//       // model: this.model.serialize(),
//       // prompt: this.prompt.serialize(),
//       // humanPrefix: this.humanPrefix,
//       // aiPrefix: this.aiPrefix,
//     };
//   }
//
//   _chainType() {
//     return "chat-qa" as const;
//   }
// }
//
// export interface SerializedChatQAChain {
//   _type: "chat-qa";
// }
