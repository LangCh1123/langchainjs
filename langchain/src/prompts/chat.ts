// Default generic "any" values are for backwards compatibility.
// Replace with "string" when we are comfortable with a breaking change.

import {
  AIChatMessage,
  BaseChatMessage,
  BasePromptValue,
  ChatMessage,
  HumanChatMessage,
  InputValues,
  PartialValues,
  SystemChatMessage,
} from "../schema/index.js";
import {
  BasePromptTemplate,
  BasePromptTemplateInput,
  BaseStringPromptTemplate,
} from "./base.js";
import { PromptTemplate } from "./prompt.js";
import {
  SerializedChatPromptTemplate,
  SerializedMessagePromptTemplate,
} from "./serde.js";

export abstract class BaseMessagePromptTemplate<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  InputVariableName extends string = any
> {
  abstract inputVariables: InputVariableName[];

  abstract formatMessages(values: InputValues): Promise<BaseChatMessage[]>;

  serialize(): SerializedMessagePromptTemplate {
    return {
      _type: this.constructor.name,
      ...JSON.parse(JSON.stringify(this)),
    };
  }
}

export class ChatPromptValue extends BasePromptValue {
  messages: BaseChatMessage[];

  constructor(messages: BaseChatMessage[]) {
    super();
    this.messages = messages;
  }

  toString() {
    return JSON.stringify(this.messages);
  }

  toChatMessages() {
    return this.messages;
  }
}

export class MessagesPlaceholder<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  InputVariableName extends string = any
> extends BaseMessagePromptTemplate<InputVariableName> {
  variableName: InputVariableName;

  constructor(variableName: InputVariableName) {
    super();
    this.variableName = variableName;
  }

  get inputVariables() {
    return [this.variableName];
  }

  formatMessages(
    values: InputValues<InputVariableName>
  ): Promise<BaseChatMessage[]> {
    return Promise.resolve(values[this.variableName] as BaseChatMessage[]);
  }
}

export abstract class BaseMessageStringPromptTemplate<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  InputVariableName extends string = any
> extends BaseMessagePromptTemplate<InputVariableName> {
  prompt: BaseStringPromptTemplate<InputVariableName, string>;

  protected constructor(
    prompt: BaseStringPromptTemplate<InputVariableName, string>
  ) {
    super();
    this.prompt = prompt;
  }

  get inputVariables() {
    return this.prompt.inputVariables;
  }

  abstract format(
    values: InputValues<InputVariableName>
  ): Promise<BaseChatMessage>;

  async formatMessages(
    values: InputValues<InputVariableName>
  ): Promise<BaseChatMessage[]> {
    return [await this.format(values)];
  }
}

export abstract class BaseChatPromptTemplate<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  InputVariableName extends string = any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  PartialVariableName extends string = any
> extends BasePromptTemplate<InputVariableName, PartialVariableName> {
  constructor(
    input: BasePromptTemplateInput<InputVariableName, PartialVariableName>
  ) {
    super(input);
  }

  abstract formatMessages(
    values: InputValues<InputVariableName>
  ): Promise<BaseChatMessage[]>;

  async format(values: InputValues<InputVariableName>): Promise<string> {
    return (await this.formatPromptValue(values)).toString();
  }

  async formatPromptValue(
    values: InputValues<InputVariableName>
  ): Promise<BasePromptValue> {
    const resultMessages = await this.formatMessages(values);
    return new ChatPromptValue(resultMessages);
  }
}

export class ChatMessagePromptTemplate<
  InputVariableName extends string
> extends BaseMessageStringPromptTemplate<InputVariableName> {
  role: string;

  async format(
    values: InputValues<InputVariableName>
  ): Promise<BaseChatMessage> {
    return new ChatMessage(await this.prompt.format(values), this.role);
  }

  constructor(
    prompt: BaseStringPromptTemplate<InputVariableName, string>,
    role: string
  ) {
    super(prompt);
    this.role = role;
  }

  static fromTemplate(template: string, role: string) {
    return new this(PromptTemplate.fromTemplate(template), role);
  }
}

export class HumanMessagePromptTemplate<
  InputVariableName extends string
> extends BaseMessageStringPromptTemplate<InputVariableName> {
  async format(
    values: InputValues<InputVariableName>
  ): Promise<BaseChatMessage> {
    return new HumanChatMessage(await this.prompt.format(values));
  }

  constructor(prompt: BaseStringPromptTemplate<InputVariableName, string>) {
    super(prompt);
  }

  static fromTemplate(template: string) {
    return new this(PromptTemplate.fromTemplate(template));
  }
}

export class AIMessagePromptTemplate<
  InputVariableName extends string
> extends BaseMessageStringPromptTemplate<InputVariableName> {
  async format(
    values: InputValues<InputVariableName>
  ): Promise<BaseChatMessage> {
    return new AIChatMessage(await this.prompt.format(values));
  }

  constructor(prompt: BaseStringPromptTemplate<InputVariableName, string>) {
    super(prompt);
  }

  static fromTemplate(template: string) {
    return new this(PromptTemplate.fromTemplate(template));
  }
}

export class SystemMessagePromptTemplate<
  InputVariableName extends string
> extends BaseMessageStringPromptTemplate<InputVariableName> {
  async format(
    values: InputValues<InputVariableName>
  ): Promise<BaseChatMessage> {
    return new SystemChatMessage(await this.prompt.format(values));
  }

  constructor(prompt: BaseStringPromptTemplate<InputVariableName, string>) {
    super(prompt);
  }

  static fromTemplate(template: string) {
    return new this(PromptTemplate.fromTemplate(template));
  }
}

export interface ChatPromptTemplateInput<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  InputVariableName extends string = any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  PartialVariableName extends string = any
> extends BasePromptTemplateInput<InputVariableName, PartialVariableName> {
  /**
   * The prompt messages
   */
  promptMessages: BaseMessagePromptTemplate[];

  /**
   * Whether to try validating the template on initialization
   *
   * @defaultValue `true`
   */
  validateTemplate?: boolean;
}

export class ChatPromptTemplate<
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    InputVariableName extends string = any,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    PartialVariableName extends string = any
  >
  extends BaseChatPromptTemplate<InputVariableName, PartialVariableName>
  implements ChatPromptTemplateInput<InputVariableName, PartialVariableName>
{
  promptMessages: BaseMessagePromptTemplate[];

  validateTemplate = true;

  constructor(
    input: ChatPromptTemplateInput<InputVariableName, PartialVariableName>
  ) {
    super(input);
    Object.assign(this, input);

    if (this.validateTemplate) {
      const inputVariablesMessages = new Set<string>();
      for (const promptMessage of this.promptMessages) {
        for (const inputVariable of promptMessage.inputVariables) {
          inputVariablesMessages.add(inputVariable);
        }
      }

      const totalInputVariables = this.inputVariables as string[];
      const inputVariablesInstance = new Set(
        this.partialVariables
          ? totalInputVariables.concat(Object.keys(this.partialVariables))
          : totalInputVariables
      );
      const difference = new Set(
        [...inputVariablesInstance].filter(
          (x) => !inputVariablesMessages.has(x as InputVariableName)
        )
      );
      if (difference.size > 0) {
        throw new Error(
          `Input variables \`${[
            ...difference,
          ]}\` are not used in any of the prompt messages.`
        );
      }
      const otherDifference = new Set(
        [...inputVariablesMessages].filter(
          (x) => !inputVariablesInstance.has(x)
        )
      );
      if (otherDifference.size > 0) {
        throw new Error(
          `Input variables \`${[
            ...otherDifference,
          ]}\` are used in prompt messages but not in the prompt template.`
        );
      }
    }
  }

  _getPromptType(): "chat" {
    return "chat";
  }

  async formatMessages(
    values: InputValues<InputVariableName>
  ): Promise<BaseChatMessage[]> {
    const allValues = await this.mergePartialAndUserVariables(values);

    let resultMessages: BaseChatMessage[] = [];

    for (const promptMessage of this.promptMessages) {
      const inputValues = promptMessage.inputVariables.reduce(
        (acc, inputVariable) => {
          if (!(inputVariable in allValues)) {
            throw new Error(
              `Missing value for input variable \`${inputVariable}\``
            );
          }
          acc[inputVariable] = allValues[inputVariable as InputVariableName];
          return acc;
        },
        {} as InputValues
      );
      const message = await promptMessage.formatMessages(inputValues);
      resultMessages = resultMessages.concat(message);
    }
    return resultMessages;
  }

  serialize(): SerializedChatPromptTemplate {
    if (this.outputParser !== undefined) {
      throw new Error(
        "ChatPromptTemplate cannot be serialized if outputParser is set"
      );
    }
    return {
      input_variables: this.inputVariables,
      prompt_messages: this.promptMessages.map((m) => m.serialize()),
    };
  }

  async partial<NewPartialVariableName extends string>(
    values: PartialValues<NewPartialVariableName>
  ) {
    // This is implemented in a way it doesn't require making
    // BaseMessagePromptTemplate aware of .partial()
    const newInputVariables = this.inputVariables.filter(
      (iv) => !(iv in values)
    ) as Exclude<InputVariableName, NewPartialVariableName>[];
    const newPartialVariables = {
      ...(this.partialVariables ?? {}),
      ...values,
    } as PartialValues<PartialVariableName | NewPartialVariableName>;
    const promptDict = {
      ...this,
      inputVariables: newInputVariables,
      partialVariables: newPartialVariables,
    };
    return new ChatPromptTemplate(promptDict);
  }

  static fromPromptMessages(
    promptMessages: (
      | BaseMessagePromptTemplate<string>
      | ChatPromptTemplate<string, string>
    )[]
  ): ChatPromptTemplate<string, string> {
    const flattenedMessages = promptMessages.reduce(
      (acc, promptMessage) =>
        acc.concat(
          // eslint-disable-next-line no-instanceof/no-instanceof
          promptMessage instanceof ChatPromptTemplate
            ? promptMessage.promptMessages
            : [promptMessage]
        ),
      [] as BaseMessagePromptTemplate[]
    );
    const flattenedPartialVariables = promptMessages.reduce(
      (acc, promptMessage) =>
        // eslint-disable-next-line no-instanceof/no-instanceof
        promptMessage instanceof ChatPromptTemplate
          ? Object.assign(acc, promptMessage.partialVariables)
          : acc,
      Object.create(null) as PartialValues
    );

    const inputVariables = new Set<string>();
    for (const promptMessage of flattenedMessages) {
      for (const inputVariable of promptMessage.inputVariables) {
        if (inputVariable in flattenedPartialVariables) {
          continue;
        }
        inputVariables.add(inputVariable);
      }
    }
    return new ChatPromptTemplate({
      inputVariables: [...inputVariables],
      promptMessages: flattenedMessages,
      partialVariables: flattenedPartialVariables,
    });
  }
}
