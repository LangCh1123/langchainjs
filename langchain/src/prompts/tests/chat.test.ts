import { expect, test } from "@jest/globals";
import { ChatPromptTemplate } from "../chat.js";
import { PromptTemplate } from "../prompt.js";

test("Test format", async () => {
  const systemPrompt = new PromptTemplate({
    template: "Here's some context: {context}",
    inputVariables: ["context"],
  });
  const userPrompt = new PromptTemplate({
    template: "Hello {foo}, I'm {bar}",
    inputVariables: ["foo", "bar"],
  });
  const chatPrompt = new ChatPromptTemplate({
    promptMessages: [
      {
        role: "system",
        message: systemPrompt,
      },
      {
        role: "user",
        message: userPrompt,
      },
    ],
    inputVariables: ["context", "foo", "bar"],
  });
  const messages = await chatPrompt.formatChat({
    context: "This is a context",
    foo: "Foo",
    bar: "Bar",
  });
  expect(messages).toEqual([
    {
      role: "system",
      text: "Here's some context: This is a context",
    },
    {
      role: "user",
      text: "Hello Foo, I'm Bar",
    },
  ]);
});

test("Test format with invalid input values", async () => {
  const systemPrompt = new PromptTemplate({
    template: "Here's some context: {context}",
    inputVariables: ["context"],
  });
  const userPrompt = new PromptTemplate({
    template: "Hello {foo}, I'm {bar}",
    inputVariables: ["foo", "bar"],
  });
  const chatPrompt = new ChatPromptTemplate({
    promptMessages: [
      {
        role: "system",
        message: systemPrompt,
      },
      {
        role: "user",
        message: userPrompt,
      },
    ],
    inputVariables: ["context", "foo", "bar"],
  });
  await expect(
    chatPrompt.formatChat({
      context: "This is a context",
      foo: "Foo",
    })
  ).rejects.toThrow("Missing value for input variable `bar`");
});

test("Test format with invalid input variables", async () => {
  const systemPrompt = new PromptTemplate({
    template: "Here's some context: {context}",
    inputVariables: ["context"],
  });
  const userPrompt = new PromptTemplate({
    template: "Hello {foo}, I'm {bar}",
    inputVariables: ["foo", "bar"],
  });
  expect(
    () =>
      new ChatPromptTemplate({
        promptMessages: [
          {
            role: "system",
            message: systemPrompt,
          },
          {
            role: "user",
            message: userPrompt,
          },
        ],
        inputVariables: ["context", "foo", "bar", "baz"],
      })
  ).toThrow(
    "Input variables `baz` are not used in any of the prompt messages."
  );

  expect(
    () =>
      new ChatPromptTemplate({
        promptMessages: [
          {
            role: "system",
            message: systemPrompt,
          },
          {
            role: "user",
            message: userPrompt,
          },
        ],
        inputVariables: ["context", "foo"],
      })
  ).toThrow(
    "Input variables `bar` are used in prompt messages but not in the prompt template."
  );
});

test("Test fromPromptMessages", async () => {
  const systemPrompt = new PromptTemplate({
    template: "Here's some context: {context}",
    inputVariables: ["context"],
  });
  const userPrompt = new PromptTemplate({
    template: "Hello {foo}, I'm {bar}",
    inputVariables: ["foo", "bar"],
  });
  const chatPrompt = ChatPromptTemplate.fromPromptMessages([
    {
      role: "system",
      message: systemPrompt,
    },
    {
      role: "user",
      message: userPrompt,
    },
  ]);
  expect(chatPrompt.inputVariables).toEqual(["context", "foo", "bar"]);
  const messages = await chatPrompt.formatChat({
    context: "This is a context",
    foo: "Foo",
    bar: "Bar",
  });
  expect(messages).toEqual([
    {
      role: "system",
      text: "Here's some context: This is a context",
    },
    {
      role: "user",
      text: "Hello Foo, I'm Bar",
    },
  ]);
});
