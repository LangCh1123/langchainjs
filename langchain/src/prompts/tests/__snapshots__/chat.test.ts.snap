// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test serialize 1`] = `
{
  "input_variables": [
    "context",
    "foo",
    "bar",
  ],
  "prompt_messages": [
    {
      "_type": "SystemMessagePromptTemplate",
      "prompt": {
        "inputSchema": {
          "_cached": null,
          "_def": {
            "catchall": {
              "_def": {
                "typeName": "ZodNever",
              },
            },
            "typeName": "ZodObject",
            "unknownKeys": "strip",
          },
        },
        "inputVariables": [
          "context",
        ],
        "template": "Here's some context: {context}",
        "templateFormat": "f-string",
        "validateTemplate": true,
      },
    },
    {
      "_type": "HumanMessagePromptTemplate",
      "prompt": {
        "inputSchema": {
          "_cached": null,
          "_def": {
            "catchall": {
              "_def": {
                "typeName": "ZodNever",
              },
            },
            "typeName": "ZodObject",
            "unknownKeys": "strip",
          },
        },
        "inputVariables": [
          "foo",
          "bar",
          "context",
        ],
        "template": "Hello {foo}, I'm {bar}. Thanks for the {context}",
        "templateFormat": "f-string",
        "validateTemplate": true,
      },
    },
    {
      "_type": "AIMessagePromptTemplate",
      "prompt": {
        "inputSchema": {
          "_cached": null,
          "_def": {
            "catchall": {
              "_def": {
                "typeName": "ZodNever",
              },
            },
            "typeName": "ZodObject",
            "unknownKeys": "strip",
          },
        },
        "inputVariables": [
          "foo",
          "bar",
        ],
        "template": "I'm an AI. I'm {foo}. I'm {bar}.",
        "templateFormat": "f-string",
        "validateTemplate": true,
      },
    },
    {
      "_type": "ChatMessagePromptTemplate",
      "prompt": {
        "inputSchema": {
          "_cached": null,
          "_def": {
            "catchall": {
              "_def": {
                "typeName": "ZodNever",
              },
            },
            "typeName": "ZodObject",
            "unknownKeys": "strip",
          },
        },
        "inputVariables": [
          "foo",
          "bar",
        ],
        "template": "I'm a generic message. I'm {foo}. I'm {bar}.",
        "templateFormat": "f-string",
        "validateTemplate": true,
      },
      "role": "test",
    },
  ],
}
`;
